enhancements/machine-config/mco-state-reporting.md

```diff
@@ -0,0 +1,453 @@
+see-also:
+replaces:
+superseded-by:
+---
+
+# Track MCO State Related to Upgrades and Everyday Operations.
+
+## Summary
+
+This enhancement describes how Nodes and their upgrade processes should be aggregated by the MCO into a user facing object. Alongside this, MCO component progression should be aggregated into the machineconfiguration operator/v1 object. The goal here is to allow customers and the MCO team to decipher the our processes in a more verbose way, speeding up the debugging process and allowing for better customer engagement.
+
+## Motivation
+
+The MCO manages node upgrades but since we do not own the object or store much of their data in other ways, much of what occurs during an upgrade
+is simply a black box operation that we currently report as "Updating" or "Updated". Users can debug into a specific node or look into the node spec for some of this information
+but most of it simply lives in the MCO code rather than in data structures. We want to put these abstract "phases" of node operations as triggered by the MCO into a concrete data structure. 
+
+The MCO also has "day to day" operations on its MachineConfigController, MachineConfigDaemon, and MachineConfigOperator pods. Tracking these daily progression statuses is just as important as the upgrade status. However, this information does not belong in the upgrade-progression datatype. Currently, only the MCO team has the skillset to dive into the code and figure out what "phase" of the MCO causes a non-upgrade related error. The goal of this work is to make that information accessible and understandable to an everyday audience.
+
+### User Stories
+
+* As a <role>, I want to <take some action> so that I can <accomplish a
+goal>.
+
+### Goals
+
+* Make a MachineConfigState type that succinctly holds the upgrade data.
+* Have API load be as minimal as possible but augment the proper objects as needed.
+* Add State reporting to the operator machineconfiguration object, and add the MCO logic for updating these structs.
+* aggregate as much MCO related data into easily accessible places as possible.
+
+### Non-Goals
+
+* Make existing/new API objects difficult to understand.
+
+## Proposal
+
+Create a mechanism inside of the MCO to update the new MachineConfigState API type
+with data about a Node's progress during an upgrade and any errors that occur during those
+processes. Also use this mechanism to update the machineconfiguration object with everyday information.
+
+The data created by this MachineConfigState would look like the following:
+
+```console
+$ oc get mcs
+NAME             HEALTH    PROGRESS   NODE                          ACTION      ERROR
+upgrade-master   Healthy   Ready      ip-10-0-35-46.ec2.internal    NodeReady   
+upgrade-worker   Healthy   Ready      ip-10-0-21-151.ec2.internal   NodeReady   
+```
+
+Where the upgrade-worker and upgrade-master contain the nodes in each respective pool. The itemized entries contain the
```

Comment:
Wondering if we should keep the Name consistent with pool name for easier mapping? i.e Just use master and worker, <custom-pools>